Mathlib/CategoryTheory/Shift/SingleFunctors.lean

```diff
@@ -0,0 +1,220 @@
+/-
+Copyright (c) 2024 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.CategoryTheory.Shift.CommShift
+
+/-!
+# Functors from a category to a category with a shift
+
+Given a category `C`, and a category `D` equipped with a shift by a monoid `A`,
+we define a structure `SingleFunctors C D A` which contains the data of
+functors `functor a : C ⥤ D` for all `a : A` and isomorphisms
+`shiftIso n a a' h : functor a' ⋙ shiftFunctor D n ≅ functor a`
+whenever `n + a = a'`. These isomorphisms should satisfy certain compatibilities
+with respect to the shift on `D`.
+
+This notion is similar to `Functor.ShiftSequence` which can be used in order to
+attach shifted versions of a homological functor `D ⥤ C` with `D` a
+triangulated category and `C` an abelian category. However, the definition
+`SingleFunctors` is for functors in the other direction: it is meant to
+ease the formalization of the compatibilities with shifts of the
+functors `C ⥤ CochainComplex C ℤ` (or `C ⥤ DerivedCategory C` (TODO))
+which sends an object `X : C` to a complex where `X` sits in a single degree.
+
+-/
+
+open CategoryTheory Category ZeroObject Limits
+
+variable (C D E E' : Type*) [Category C] [Category D] [Category E]
+  (A : Type*) [AddMonoid A] [HasShift D A] [HasShift E A]
+
+namespace CategoryTheory
+
+/-- The type of families of functors `A → C ⥤ D` which are compatible with
+the shift by `A` on the category `D`. -/
+structure SingleFunctors where
+  /-- a family of functors `C ⥤ D` indexed by the elements of the additive monoid `A` -/
+  functor (a : A) : C ⥤ D
+  /-- the isomorphism `functor a' ⋙ shiftFunctor D n ≅ functor a` when `n + a = a'` -/
+  shiftIso (n a a' : A) (ha' : n + a = a') : functor a' ⋙ shiftFunctor D n ≅ functor a
+  shiftIso_zero (a : A) :
+    shiftIso 0 a a (zero_add a) = isoWhiskerLeft _ (shiftFunctorZero D A)
+  shiftIso_add (n m a a' a'' : A) (ha' : n + a = a') (ha'' : m + a' = a'') :
+    shiftIso (m + n) a a'' (by rw [add_assoc, ha', ha'']) =
+      isoWhiskerLeft _ (shiftFunctorAdd D m n) ≪≫ (Functor.associator _ _ _).symm ≪≫
+        isoWhiskerRight (shiftIso m a' a'' ha'') _ ≪≫ shiftIso n a a' ha'
```

Comment:
```suggestion
  shiftIso (n a a' : A) (ha' : n + a = a') : functor a' ⋙ shiftFunctor D n ≅ functor a
  /-- `shiftIso 0` is the identity up to `shiftFunctorZero`. -/
  shiftIso_zero (a : A) :
    shiftIso 0 a a (zero_add a) = isoWhiskerLeft _ (shiftFunctorZero D A)
  /-- `shiftIso (m + n)` is the composition of `shiftIso n` with `shiftIso m` up
  to `shiftFunctorAdd`. -/
  shiftIso_add (n m a a' a'' : A) (ha' : n + a = a') (ha'' : m + a' = a'') :
    shiftIso (m + n) a a'' (by rw [add_assoc, ha', ha'']) =
      isoWhiskerLeft _ (shiftFunctorAdd D m n) ≪≫ (Functor.associator _ _ _).symm ≪≫
        isoWhiskerRight (shiftIso m a' a'' ha'') _ ≪≫ shiftIso n a a' ha'
```
Maybe some description of the compatibility conditions?

Reply:
I have added docstrings.